src/swap/defi/xrpDex.ts

```diff
@@ -0,0 +1,275 @@
+import { round } from 'biggystring'
+import { asArray, asNumber, asObject, asOptional, asString } from 'cleaners'
+import {
+  EdgeCorePluginOptions,
+  EdgeSwapInfo,
+  EdgeSwapPlugin,
+  EdgeSwapQuote,
+  EdgeSwapRequest,
+  EdgeTxSwap,
+  SwapCurrencyError
+} from 'edge-core-js/types'
+import { Client } from 'xrpl'
+
+import { makeSwapPluginQuote, SwapOrder } from '../../swap-helpers'
+import {
+  convertRequest,
+  fetchInfo,
+  getAddress,
+  promiseWithTimeout,
+  shuffleArray
+} from '../../util/utils'
+import { EdgeSwapRequestPlugin } from '../types'
+import { getBuyQuote, getSellQuote } from './xrp/xrpDexHelpers'
+import { MakeTxParams } from './xrp/xrpDexTypes'
+
+const pluginId = 'xrpdex'
+const swapInfo: EdgeSwapInfo = {
+  pluginId,
+  isDex: true,
+  displayName: 'XRP DEX',
+  supportEmail: 'support@edge.app'
+}
+
+export const asInitOptions = asObject({
+  appId: asOptional(asString, 'edge')
+})
+
+const EXPIRATION_MS = 1000 * 60
+const DEX_MAX_FULLFILLMENT_TIME_S = 5 * 60 // 5 mins
+const RIPPLE_SERVERS_DEFAULT = ['wss://s2.ripple.com']
+const EXCHANGE_INFO_UPDATE_FREQ_MS = 60000
+const VOLATILITY_SPREAD_DEFAULT = 0.0075
+
+export const asExchangeInfo = asObject({
+  swap: asObject({
+    plugins: asObject({
+      xrpdex: asObject({
+        volatilitySpread: asNumber,
+        rippleServers: asOptional(asArray(asString))
+      })
+    })
+  })
+})
+
+type ExchangeInfo = ReturnType<typeof asExchangeInfo>
+
+let exchangeInfo: ExchangeInfo | undefined
+let exchangeInfoLastUpdate: number = 0
+
+export function makeXrpDexPlugin(opts: EdgeCorePluginOptions): EdgeSwapPlugin {
+  const { io, log } = opts
+  const { fetchCors = io.fetch } = io
+  const { appId } = asInitOptions(opts.initOptions)
+
+  const fetchSwapQuoteInner = async (
+    request: EdgeSwapRequestPlugin
+  ): Promise<SwapOrder> => {
+    const {
+      fromCurrencyCode,
+      fromTokenId,
+      toCurrencyCode,
+      toTokenId,
+      nativeAmount,
+      fromWallet,
+      toWallet,
+      quoteFor
+    } = request
+
+    // Only support ripple wallets
+    if (
+      fromWallet.currencyInfo.pluginId !== 'ripple' ||
+      toWallet.currencyInfo.pluginId !== 'ripple'
+    ) {
+      throw new SwapCurrencyError(swapInfo, fromCurrencyCode, toCurrencyCode)
```

Comment:
Pass the `request` object to `SwapCurrencyError`. Passing the currency codes is deprecated.